python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+        :param stop_upstream:   If true then all upstream tasks will also be stopped
+        :param stop_downstream: If true then all downstream tasks will also be stopped
+        """
+        task = self._tasks_by_id.get(task_id)
+        if task is None:
+            return
+
+        self._log("Stopping Task %s..." % task)
+        self._stop_tasks([task], stop_upstream, stop_downstream)
+        self._log(" > Task %s stopped!" % task)
+
+    def stop_task_group(self, group, stop_upstream=True, stop_downstream=True):
+        """
+        Stop all tasks in the specified group from running.  If any tasks are already running then the will 
+        complete but their completion/failure signals will be ignored.
+
+        :param group:           The task group to stop
+        :param stop_upstream:   If true then all upstream tasks will also be stopped
+        :param stop_downstream: If true then all downstream tasks will also be stopped
+        """
+        task_ids = self._group_task_map.get(group)
+        if task_ids is None:
+            return
+
+        self._log("Stopping Task group %s..." % group)
+
+        tasks_to_stop = []
+        for task_id in task_ids:
+            task = self._tasks_by_id.get(task_id)
+            if task:
+                tasks_to_stop.append(task)
+        del self._group_task_map[group]
+        self._stop_tasks(tasks_to_stop, stop_upstream, stop_downstream)
+
+        self._log(" > Task group %s stopped!" % group)
+
+    def stop_all_tasks(self):
+        """
+        Stop all currently queued or running tasks.  If any tasks are already running then the will 
+        complete but their completion/failure signals will be ignored.
+        """
+        self._log("Stopping all tasks...")
+
+        # we just need to clear all the lookups:
+        self._running_tasks = {}
+        self._pending_tasks_by_priority = {}
+        self._tasks_by_id = {}
+        self._group_task_map = {}
+        self._upstream_task_map = {}
+        self._downstream_task_map = {}
+
+        self._log(" > All tasks stopped!")
+
+    def _stop_tasks(self, tasks_to_stop, stop_upstream, stop_downstream):
+        """
+        Stop the specified list of tasks
+
+        :param tasks_to_stop:   A list of tasks to stop
+        :param stop_upstream:   If true then all upstream tasks will also be stopped
+        :param stop_downstream: If true then all downstream tasks will also be stopped
+        """
+        if not tasks_to_stop:
+            return
+
+        # copy the task list as we'll be modifying it:
+        tasks_to_stop = list(tasks_to_stop)
+        # and make sure we only stop each task once!!
+        stopped_task_ids = set([task.uid for task in tasks_to_stop])
+
+        while tasks_to_stop:
+            task_to_stop = tasks_to_stop.pop(0)
+
+            # get the up & downstream tasks to also stop depending on the flags:
+            if stop_upstream and task_to_stop.uid in self._upstream_task_map:
```

Comment:
I meant it as "8 lines of copy-pasted code with only variable name changes should be refactored and turned into a method so that maintaining them would be easier"